File: src/cluster/cluster.ts
/**
 * Graph + Louvain + partition-to-target. Port of analysis/tabamnesty/cluster.py.
 *
 * Louvain, not connected components and not single-link: both chain, so one weak bridge merges
 * two real projects. Louvain optimises modularity and a lone bridge does not survive that.
 */
import Graph from 'graphology';
import louvain from 'graphology-communities-louvain';
import { affinity, Context, eligible, signalVector } from './affinity';
import {
  type Betas,
  BETAS,
  MAX_COMMUNITY,
  RESOLUTION_ITERS,
  SIGNALS,
  type SignalVector,
  TARGET_HI,
  TARGET_LO,
  W_MIN,
} from './params';
import type { Partition, TabTrace } from './types';

export type PairKey = `${string}|${string}`;
export const pairKey = (a: string, b: string): PairKey => `${a}|${b}`;

/** Deterministic PRNG so Louvain gives the same answer for the same input (mulberry32). */
function seededRng(seed = 0): () => number {
  let s = seed >>> 0;
  return () => {
    s = (s + 0x6d2b79f5) >>> 0;
    let t = s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/** Every unordered pair's signal vector, keyed "a|b" in input order. O(n²). */
export function pairSignals(traces: TabTrace[]): { ctx: Context; vecs: Map<PairKey, SignalVector> } {
  const ctx = new Context(traces);
  const vecs = new Map<PairKey, SignalVector>();
  for (let i = 0; i < traces.length; i++) {
    for (let j = i + 1; j < traces.length; j++) {
      vecs.set(pairKey(traces[i]!.traceId, traces[j]!.traceId), signalVector(ctx, traces[i]!, traces[j]!));
    }
  }
  return { ctx, vecs };
}

export function buildGraph(
  traces: TabTrace[],
  betas: Betas,
  wMin = W_MIN,
  vecs?: Map<PairKey, SignalVector>,
): Graph {
  vecs ??= pairSignals(traces).vecs;
  const g = new Graph({ type: 'undirected' });
  for (const t of traces) g.addNode(t.traceId);
  for (const [key, vec] of vecs) {
    const w = affinity(vec, betas);
    if (w >= wMin) {
      const [a, b] = key.split('|') as [string, string];
      g.addEdge(a, b, { weight: w });
    }
  }
  return g;
}

function runLouvain(g: Graph, resolution: number): string[][] {
  if (g.size === 0) return g.nodes().map((n) => [n]);
  const mapping = louvain(g, { resolution, getEdgeWeight: 'weight', rng: seededRng(0) });
  const byComm = new Map<number, string[]>();
  for (const n of g.nodes()) {
    const c = mapping[n]!;
    if (!byComm.has(c)) byComm.set(c, []);
    byComm.get(c)!.push(n);
  }
  return [...byComm.values()];
}

/**
 * Binary-search Louvain's resolution (log scale) until the number of real communities
 * (size >= 2) lands in [lo, hi]. Group count is a product guarantee, not an algorithm output.
 */
export function partitionToTarget(
  g: Graph,
  lo = TARGET_LO,
  hi = TARGET_HI,
  iters = RESOLUTION_ITERS,
): { communities: string[][]; resolution: number } {
  let loR = Math.log(0.02);
  let hiR = Math.log(20);
  let best: string[][] | null = null;
  let bestRes = 1;
  let bestGap = Infinity;
  for (let i = 0; i < iters; i++) {
    const mid = (loR + hiR) / 2;
    const res = Math.exp(mid);
    const comms = runLouvain(g, res);
    const n = comms.filter((c) => c.length >= 2).length;
    const gap = n >= lo && n <= hi ? 0 : Math.min(Math.abs(n - lo), Math.abs(n - hi));
    if (gap < bestGap || (gap === bestGap && Math.abs(mid) < Math.abs(Math.log(bestRes)))) {
      best = comms;
      bestRes = res;
      bestGap = gap;
    }
    if (gap === 0) break;
    if (n > hi) hiR = mid; // too many groups -> lower resolution
    else loR = mid;
  }
  return { communities: best!, resolution: bestRes };
}

export function splitLarge(g: Graph, comms: string[][], maxSize = MAX_COMMUNITY): string[][] {
  const out: string[][] = [];
  for (const c of comms) {
    if (c.length <= maxSize) {
      out.push(c);
      continue;
    }
    const sub = new Graph({ type: 'undirected' });
    for (const n of c) sub.addNode(n);
    for (const n of c) {
      g.forEachEdge(n, (_e, attrs, s, t) => {
        if (sub.hasNode(s) && sub.hasNode(t) && !sub.hasEdge(s, t)) sub.addEdge(s, t, attrs);
      });
    }
    const parts = runLouvain(sub, 1);
    if (parts.length <= 1) out.push(c); // Louvain will not split it; keep rather than force
    else out.push(...splitLarge(g, parts, maxSize));
  }
  return out;
}

export interface ClusterOptions {
  betas?: Betas;
  wMin?: number;
  lo?: number;
  hi?: number;
  vecs?: Map<PairKey, SignalVector>;
}

export interface ClusterResult extends Partition {
  resolution: number;
  /** pinned / ambient / non-http / closed — never scored as ours */
  excluded: string[];
}

export function cluster(traces: TabTrace[], opts: ClusterOptions = {}): ClusterResult {
  const betas = opts.betas ?? BETAS;
  const { kept, excluded } = eligible(traces);
  const excludedIds = excluded.map((t) => t.traceId);
  if (kept.length === 0) return { communities: [], looseEnds: [], resolution: 1, excluded: excludedIds };
  const g = buildGraph(kept, betas, opts.wMin ?? W_MIN, opts.vecs);
  const { communities: raw, resolution } = partitionToTarget(g, opts.lo, opts.hi);
  const comms = splitLarge(g, raw);
  const order = new Map(kept.map((t, i) => [t.traceId, i]));
  const ord = (id: string) => order.get(id) ?? Infinity;
  const real = comms
    .filter((c) => c.length >= 2)
    .map((c) => [...c].sort((a, b) => ord(a) - ord(b)))
    .sort((a, b) => b.length - a.length || ord(a[0]!) - ord(b[0]!));
  const looseEnds = comms
    .filter((c) => c.length < 2)
    .map((c) => c[0]!)
    .sort((a, b) => ord(a) - ord(b));
  return {
    communities: real.map((traceIds, id) => ({ id, traceIds })),
    looseEnds,
    resolution,
    excluded: excludedIds,
  };
}

/** Every beta zeroed one at a time, plus the S1+S2+S8 gate condition together. */
export function ablationBetas(betas: Betas = BETAS): Record<string, Betas> {
  const runs: Record<string, Betas> = { baseline: { ...betas } };
  for (const s of SIGNALS) runs[`no_${s}`] = { ...betas, [s]: 0 };
  runs.no_S1_S2_S8 = { ...betas, S1_lineage: 0, S2_temporal: 0, S8_coactive: 0 };
  return runs;
}

File: analysis/tabamnesty/cluster.py
"""Graph + Louvain + partition-to-target. The whole clusterer, end to end.

Louvain, not connected components and not single-link: both chain, so one weak bridge merges
two real projects. Louvain optimises modularity and a lone bridge does not survive that."""
from __future__ import annotations

import itertools
import math
from dataclasses import dataclass, field

import networkx as nx

from .config import (MAX_COMMUNITY, RESOLUTION_ITERS, SIGNALS, TARGET_HI, TARGET_LO, W_MIN,
                     load_betas)
from .louvain import louvain
from .signals import Context, affinity, eligible, signal_vector
from .traces import Trace


@dataclass
class Partition:
    communities: list[list[str]]
    loose_ends: list[str]
    resolution: float = 1.0
    excluded: list[str] = field(default_factory=list)  # pinned / ambient / non-http, never scored as ours

    def to_json(self, source: str, betas: dict[str, float]) -> dict:
        return {
            "schemaVersion": 1,
            "source": source,
            "betas": betas,
            "resolution": self.resolution,
            "communities": [{"id": i, "traceIds": c} for i, c in enumerate(self.communities)],
            "looseEnds": self.loose_ends,
            "excluded": self.excluded,
        }


def pair_signals(traces: list[Trace]) -> tuple[Context, dict[tuple[str, str], dict[str, float]]]:
    """Every unordered pair's signal vector. O(n²); fine for a few hundred tabs."""
    ctx = Context(traces)
    vecs = {}
    for a, b in itertools.combinations(traces, 2):
        vecs[(a["traceId"], b["traceId"])] = signal_vector(ctx, a, b)
    return ctx, vecs


def build_graph(traces: list[Trace], betas: dict[str, float], w_min: float = W_MIN,
                vecs: dict[tuple[str, str], dict[str, float]] | None = None) -> nx.Graph:
    if vecs is None:
        _, vecs = pair_signals(traces)
    g = nx.Graph()
    g.add_nodes_from(t["traceId"] for t in traces)
    for (a, b), vec in vecs.items():
        w = affinity(vec, betas)
        if w >= w_min:
            g.add_edge(a, b, weight=w)
    return g


def _louvain(g: nx.Graph, resolution: float) -> list[set[str]]:
    """Our deterministic Louvain (louvain.py), not networkx's — see that file for why."""
    edges = [(a, b, d["weight"]) for a, b, d in g.edges(data=True)]
    return [set(c) for c in louvain(list(g.nodes), edges, resolution)]


def partition_to_target(g: nx.Graph, lo: int = TARGET_LO, hi: int = TARGET_HI,
                        iters: int = RESOLUTION_ITERS) -> tuple[list[set[str]], float]:
    """Binary-search Louvain's resolution (log scale) until the number of real communities
    (size >= 2) lands in [lo, hi]. Group count is a product guarantee, not an algorithm output."""
    lo_r, hi_r = math.log(0.02), math.log(20.0)
    best, best_res, best_gap = None, 1.0, math.inf
    for _ in range(iters):
        mid = (lo_r + hi_r) / 2
        res = math.exp(mid)
        comms = _louvain(g, res)
        n = sum(1 for c in comms if len(c) >= 2)
        gap = 0 if lo <= n <= hi else min(abs(n - lo), abs(n - hi))
        if gap < best_gap or (gap == best_gap and abs(mid) < abs(math.log(best_res))):
            best, best_res, best_gap = comms, res, gap
        if gap == 0:
            break
        if n > hi:
            hi_r = mid  # too many groups -> lower resolution
        else:
            lo_r = mid
    assert best is not None
    return best, best_res


def _induced(g: nx.Graph, nodes: set[str]) -> nx.Graph:
    """Induced subgraph with nodes in the parent graph's order. Not g.subgraph(): a subgraph view
    over a small node set iterates the *set*, so Louvain's tie-breaking — and the partition —
    would depend on PYTHONHASHSEED."""
    h = nx.Graph()
    h.add_nodes_from(n for n in g.nodes if n in nodes)
    h.add_edges_from((a, b, d) for a, b, d in g.edges(data=True) if a in nodes and b in nodes)
    return h


def split_large(g: nx.Graph, comms: list[set[str]], max_size: int = MAX_COMMUNITY) -> list[set[str]]:
    out: list[set[str]] = []
    for c in comms:
        if len(c) <= max_size:
            out.append(c)
            continue
        sub = _louvain(_induced(g, c), 1.0)
        if len(sub) <= 1:  # Louvain will not split it; keep rather than force
            out.append(c)
        else:
            out.extend(split_large(g, sub, max_size))
    return out


def cluster(traces: list[Trace], betas: dict[str, float] | None = None, w_min: float = W_MIN,
            lo: int = TARGET_LO, hi: int = TARGET_HI,
            vecs: dict[tuple[str, str], dict[str, float]] | None = None) -> Partition:
    betas = betas or load_betas()
    kept, excluded = eligible(traces)
    if not kept:
        return Partition([], [], excluded=[t["traceId"] for t in excluded])
    g = build_graph(kept, betas, w_min, vecs)
    comms, res = partition_to_target(g, lo, hi)
    comms = split_large(g, comms)
    order = {t["traceId"]: i for i, t in enumerate(kept)}
    real = sorted((sorted(c, key=order.get) for c in comms if len(c) >= 2),
                  key=lambda c: (-len(c), order[c[0]]))
    loose = sorted((next(iter(c)) for c in comms if len(c) < 2), key=order.get)
    return Partition(real, loose, res, [t["traceId"] for t in excluded])


def ablation_betas(betas: dict[str, float]) -> dict[str, dict[str, float]]:
    """Every beta zeroed one at a time, plus the S1+S2+S8 gate condition together."""
    runs = {"baseline": dict(betas)}
    for s in SIGNALS:
        runs[f"no_{s}"] = {**betas, s: 0.0}
    runs["no_S1_S2_S8"] = {**betas, "S1_lineage": 0.0, "S2_temporal": 0.0, "S8_coactive": 0.0}
    return runs
